CRM/Activity/Tokens.php

```diff
@@ -140,73 +106,97 @@ public function prefetch(TokenValueEvent $e) {
+      $row->format('text/plain')->tokens($entity, $field, $row->tokens['activity'][$realField]);
     }
     elseif (in_array($field, ['case_id'])) {
       // An activity can be linked to multiple cases so case_id is always an array.
       // We just return the first case ID for the token.
-      $row->tokens($entity, $field, is_array($activity['case_id']) ? reset($activity['case_id']) : $activity['case_id']);
+      // this weird hack might exist because apiv3 is weird &
+      $caseID = CRM_Core_DAO::singleValueQuery('SELECT case_id FROM civicrm_case_activity WHERE activity_id = %1 LIMIT 1', [1 => [$activityId, 'Integer']]);
+      $row->tokens($entity, $field, $caseID ?? '');
     }
-    elseif (array_key_exists($field, $this->customFieldTokens)) {
-      $row->tokens($entity, $field,
-        isset($activity[$field])
-          ? \CRM_Core_BAO_CustomField::displayValue($activity[$field], $field)
-          : ''
-      );
+    else {
+      parent::evaluateToken($row, $entity, $field, $prefetch);
     }
-    elseif (isset($activity[$field])) {
-      $row->tokens($entity, $field, $activity[$field]);
+  }
+
+  /**
+   * Get all the tokens supported by this processor.
+   *
+   * @return array|string[]
+   * @throws \API_Exception
+   */
+  protected function getAllTokens(): array {
+    $tokens = parent::getAllTokens();
+    if (array_key_exists('CiviCase', CRM_Core_Component::getEnabledComponents())) {
+      $tokens['case_id'] = ts('Activity Case ID');
     }
+    return $tokens;
   }
 
   /**
    * Get the basic tokens provided.
    *
    * @return array token name => token label
    */
-  protected function getBasicTokens(): array {
+  public function getBasicTokens(): array {
     if (!isset($this->basicTokens)) {
       $this->basicTokens = [
-        'activity_id' => ts('Activity ID'),
-        'activity_type' => ts('Activity Type'),
+        'id' => ts('Activity ID'),
         'subject' => ts('Activity Subject'),
         'details' => ts('Activity Details'),
         'activity_date_time' => ts('Activity Date-Time'),
         'created_date' => ts('Activity Created Date'),
         'modified_date' => ts('Activity Modified Date'),
         'activity_type_id' => ts('Activity Type ID'),
-        'status' => ts('Activity Status'),
         'status_id' => ts('Activity Status ID'),
         'location' => ts('Activity Location'),
         'duration' => ts('Activity Duration'),
-        'campaign' => ts('Activity Campaign'),
-        'campaign_id' => ts('Activity Campaign ID'),
       ];
-      if (array_key_exists('CiviCase', CRM_Core_Component::getEnabledComponents())) {
-        $this->basicTokens['case_id'] = ts('Activity Case ID');
+      if (CRM_Campaign_BAO_Campaign::isCampaignEnable()) {
+        $this->basicTokens['campaign_id'] = ts('Campaign ID');
       }
     }
     return $this->basicTokens;
   }
 
+  /**
+   * @inheritDoc
+   */
+  public function getActiveTokens(TokenValueEvent $e) {
```

Comment:
Just copied back from the trait - this overrides the parent to 

a) add in the old-style tokens and
b) do weird stuff - which is the general issue with this class